File: bindings/py/v1/my_lib/lang.py
from ._generated import loader

def format_message(name: str) -> str:
    """Formats a greeting name using the Go WASM backend (Preview 1 JSON-RPC)."""
    return loader.call("lang.format", {"name": name})

def reverse_string(s: str) -> str:
    """Reverses a string using the Go WASM backend (Preview 1 JSON-RPC)."""
    return loader.call("lang.reverse", {"s": s})

File: bindings/my_lib/wasm/v1/lang.go
package main

import (
	"encoding/json"
	"gitlab.com/ouvrage-systems/labs/wasi-polyglot-bindings-reference/pkg/text"
	"gitlab.com/ouvrage-systems/labs/wasi-polyglot-bindings-reference/tools/wasi1rpc"
)

// Register lang handlers on the RPC router
func registerLang(r *wasi1rpc.Router) {
	r.Register("lang.format", func(params json.RawMessage) (any, error) {
		var args struct {
			Name string `json:"name"`
		}
		if err := json.Unmarshal(params, &args); err != nil {
			return nil, err
		}
		return text.FormatMessage(args.Name), nil
	})

	r.Register("lang.reverse", func(params json.RawMessage) (any, error) {
		var args struct {
			S string `json:"s"`
		}
		if err := json.Unmarshal(params, &args); err != nil {
			return nil, err
		}
		return text.ReverseString(args.S), nil
	})
}
